# Ejercicios Entegrables - Grupo 18

**Grupo B -  Diciembre 2024**

- Martha Castillo

- Alexander Castiblanco

### **Ejercicio 1. Cálculo del determinante de una matriz**

a) Implementa una función, determinante recursivo, que obtenga el determinante de una matriz cuadrada utilizando la definición recursiva de Laplace.

In [89]:
import numpy as np

In [90]:
def determinante_recursivo(matriz):
    det = 0
    if len(matriz) == len(matriz[0]):
        if len(matriz) == 1:
            return matriz[0][0]
        elif len(matriz) == 2:
            return matriz[0][0] * matriz[1][1] - matriz[0][1] * matriz[1][0]
        for col in range(len(matriz)):
            submatriz = [list(fila[:col]) + list(fila[col+1:]) for fila in matriz[1:]]

            det += ((-1) ** (1*col)) * matriz[0][col] * determinante_recursivo(submatriz)
    else:
        return ('Matriz no es cuadrada')
    return det

In [91]:
A = np.array([[1, 2], [3, 4]])
B = np.array([[1, 2, 2, 1], [3, 4, 1, 2], [1, 2, 4, 3], [3, 4, 2, 1]])
C = np.array([[1, 2, 3], [4, 5, 6]])

print(determinante_recursivo(A))
print(determinante_recursivo(B))
print(determinante_recursivo(C))

-2
8
Matriz no es cuadrada


b) Si A es una matriz cuadrada n×n y triangular (superior o inferior, es decir, con entradas nulas por debajo o por encima de la diagonal, respectivamente), ¿existe alguna forma de calcular de forma directa y sencilla su determinante? Justifíquese la respuesta.

In [ ]:
# Sí, la forma de calcular el determinante de forma directa y sencilla en el caso de matrices cuadradas y
# triangulares es el producto de los elementos de la diagonal de dicha matriz.
# Ya que si por ejemlo usamos la definción vista en clase del determinante:

    #det(A) = sum_(n, j=1)(a_ij)^(1+j) * det(A_ij)

#Se puede observar que todos los elementos de dicho sumatorio son igual a 0, ya que siempre va haber un elemento o
# por debajo de la diagonal o por encima de ella; el único elemento será la diagonal.

# A continuación ilustramos dos ejemplos

In [ ]:
import math

T_inf = np.array([[1,0,0], [2,8,0], [4,9,7]])
T_sup = np.array([[1,4,2,5,6], [0,1,4,8,-1], [0,0,1,4,3], [0,0,0,1,3], [0,0,0,0,1]])

diag_inf = [1,8,7]
diag_sup = [1,1,1,1,1]

print(determinante_recursivo(T_inf))
print(determinante_recursivo(T_sup))
print(math.prod(diag_inf))
print(math.prod(diag_sup))

print(determinante_recursivo(T_inf) == math.prod(diag_inf))
print(determinante_recursivo(T_sup) == math.prod(diag_sup))

56
1
56
1
True
True


c) Determínese de forma justificada cómo alteran el determinante de una matriz n×n las dos operaciones elementales siguientes:
- Intercambiar una fila (o columna) por otra fila (o columna).
- Sumar a una fila (o columna) otra fila (o columna) multiplicada por un escalar α.

In [92]:
A = np.array([[1, 2, 3], [-1, 1, 0], [5, -2, -1]]) # matriz original
print(determinante_recursivo(A))


-12


In [93]:
# 1. Intercambiar una fila (o columna) por otra fila (o columna)

# Siendo B la nueva matriz surgida de intercambiar una fila por otra en la matriz A (que en nuestro caso será la
# matriz original), el det(B) = -det(A).
# Este cambio es debido a que, el determinante es una suma de productos alternantes de los elementos de la matriz
# Intercambiar dos filas equivale a realizar una permutación transpuesta en el índice, y una permutación impar (
# es decir, solo un cambio) se multiplica el determinante por -1, si fuese par sería por 1.

B = np.array([[-1, 1, 0],[1, 2, 3], [5, -2, -1]])  # intercambiamos la primera fila con la segunda
print(determinante_recursivo(B))

12


In [94]:
# 2. Sumar a una fila (o columna) otra fila (o columna) multiplicada por un escalar a

# Siendo C la nueva matriz surgida de sumar una fila con otra multiplicada por un escalar a, el det(C) = det(A).
# En este caso, el determinante no cambia. La propiedad de multilinealidad del determinante establece que el
# determinante de una matriz es lineal con respecto a cada fila (o columna) cuando las demás filas (o columnas)
# se mantienen constantes. Sumar a una fila otra fila multiplicada por un escalar no modifica las contribuciones
# de los productos alternantes en la expansión del determinante. Esto se debe a que la nueva fila resultante es una
# combinación lineal que conserva el determinante.

C = np.array([[1, 2, 3], [1, 5, 6], [5, -2, -1]]) # sumamos la 1º fila con la 2º, multiplicada la 1º por 2
print(determinante_recursivo(C))
D = np.array([[1, 2, 3], [-1, 1, 0], [9, 6, 11]]) # sumamos la 1º fila con la 3º, multiplicada la 1º por 4
print(determinante_recursivo(D))

-12
-12


d) Investiga sobre el método de eliminación de Gauss con pivoteo parcial e impleméntalo para escalonar una matriz (es decir, convertirla en una matriz triangular inferior) a partir de las operaciones elementales descritas en el apartado anterior.

In [95]:
# # El método de eliminación de Gauss es una técnica para resolver sistemas de ecuaciones lineales, encontrar
# inversas de matrices o calcular determinantes. Una de las variantes más comunes y estables de este método es
# el pivoteo parcial, que busca minimizar los errores numéricos al trabajar con coeficientes pequeños o muy grandes.

# El objetivo del método es transformar una matriz de coeficientes A en una matriz triangular superior U, mediante
# operaciones elementales de fila. Las operaciones permitidas son el intercambio de filas, la multiplicación de una
# fila por un escalar distinto de cero y la suma de una fila con un múltiplo de otra fila. En cada paso de
# eliminación (columna por columna), el pivoteo parcial selecciona como pivote el elemento de mayor valor absoluto
# dentro de la columna activa.

# Los pasos del método con pivoteo parcial son:
    # 1. Seleccionar el pivote: identificar el valor máximo absoluto en la columna activa (desde la fila actual
    #    hacia abajo) e intercambiar la fila que contiene este valor con la fila actual.
    # 2. Eliminar elementos debajo del Pivote: usar la fila pivote para eliminar todos los elementos debajo de ella
    #    en la columna activa, aplicando las operaciones permitidas.
    # 3. Repetir el proceso hasta obtener una matriz triangular superior.

# Las ventajas del pivoteo parcial son: la estabilidad numérica, minimiza los errores relacionados con divisiones
# por valores pequeños y evita amplificaciones de errores; y la simplicidad de implementación, el pivoteo parcial
# es más fácil de implementar que el pivoteo completo, que considera tanto filas como columnas.

In [96]:
# Ponemos la misma matriz que en el apartado anterior para implementar el método

A = np.array([[1, 2, 3], [-1, 1, 0], [5, -2, -1]]) # matriz original
print(A)

# Sumamos 1º fila y 2º
A1 = np.array([[1, 2, 3], [0, 3, 3], [5, -2, -1]])
print(A1)

# Sumamos 1º fila y 3º, multiplicada la 1º por -5
A2 = np.array([[1, 2, 3], [0, 3, 3], [0, -12, -16]])
print(A2)

# Sumamos 2º fila y 3º, multiplicada la 2º por 4
A3 = np.array([[1, 2, 3], [0, 3, 3], [0, 0, -4]])
print(A3)


# comprobamos también el apartado c) otra vez
print(determinante_recursivo(A)==determinante_recursivo(A1))
print(determinante_recursivo(A)==determinante_recursivo(A2))
print(determinante_recursivo(A)==determinante_recursivo(A3))

[[ 1  2  3]
 [-1  1  0]
 [ 5 -2 -1]]
[[ 1  2  3]
 [ 0  3  3]
 [ 5 -2 -1]]
[[  1   2   3]
 [  0   3   3]
 [  0 -12 -16]]
[[ 1  2  3]
 [ 0  3  3]
 [ 0  0 -4]]
True
True
True


e) ¿Cómo se podría calcular el determinante de una matriz haciendo beneficio de la estrategia anterior y del efecto de aplicar las operaciones elementales pertinentes? Implementa una nueva función, determinante gauss, que calcule el determinante de una matriz utilizando eliminación gaussiana.

In [98]:
def determinante_gauss(matriz):
    n = len(matriz)
    det = 1
    if n == len(matriz[0]):

        matriz = [list(fila) for fila in matriz]
        for i in range(n):
            # Buscamos el pivote máximo en la columna actual
            max_fila = i
            for k in range(i + 1, n):
                if abs(matriz[k][i]) > abs(matriz[max_fila][i]):
                    max_fila = k

            if matriz[max_fila][i] == 0:
                det = 0
            if max_fila != i:
                matriz[i], matriz[max_fila] = matriz[max_fila], matriz[i]
                det *= -1  # Cambiar el signo del determinante por el intercambio de filas

            # Hacer la eliminación hacia abajo
            for k in range(i + 1, n):
                factor = matriz[k][i] / matriz[i][i]
                for j in range(i, n):
                    matriz[k][j] -= factor * matriz[i][j]

        for i in range(n):
            det *= matriz[i][i]
    else:
        return ('Matriz no es cuadrada')

    return det

In [99]:
A = np.array([[1, 2], [3, 4]])
B = np.array([[1, 2, 2, 1], [3, 4, 1, 2], [1, 2, 4, 3], [3, 4, 2, 1]])
C = np.array([[1, 2, 3], [4, 5, 6]])

print(determinante_gauss(A))
print(determinante_gauss(B))
print(determinante_gauss(C))

-2.0
7.999999999999999
Matriz no es cuadrada


f) Obtén la complejidad computacional asociada al cálculo del determinante con la definición recursiva y con el método de eliminación de Gauss con pivoteo parcial.

In [ ]:
# Definición Recursiva

# El proceso que sigue
    # 1. Selecciona una fila para la expansión.
    # 2. Calcula los determinantes de los menores, que son submatrices (n-1)x(n-1).

# La recurrencia para este cálculo es: T(n) = n * T(n-1)
# El número de operaciones sigue una progresión geométrica: T(n) = n * (n−1) * (n−2) ... 1 = n!
# Por lo tanto, la complejidad es: O(n!)

In [ ]:
# Método de Gauss

# Su complejidad proviene de tres operaciones principales:

# 1. Selección del pivote (pivoteo parcial):
    # Encontrar el elemento de mayor valor absoluto en la columna. Esto requiere recorrer n−k elementos para la
    # columna k, y se hace para cada una de las n columnas. Complejidad: O(n^2).

# 2. Eliminación hacia abajo (hacer ceros debajo del pivote):
    # Para cada fila i>k, se eliminan los elementos de la columna actual, lo que implica operar sobre (n−k−1) filas
    # y (n−k) columnas.
    # Complejidad acumulativa: O(n^3) (por las sumas y multiplicaciones necesarias).

# 3. Producto de los elementos de la diagonal:
    # Es una operación lineal con complejidad O(n).

# La operación dominante es la eliminación hacia abajo, que define la complejidad total: O(n^3).

g) Utilizando numpy.random.rand, genera matrices cuadradas aleatorias de la forma An ∈ Rn×n, para 2 ≤ n ≤ 10, y confecciona una tabla comparativa del tiempo de ejecución asociado a cada una de las variantes siguientes, interpretando los resultados:
1. Utilizando determinante recursivo.
2. Empleando determinante gauss.
3. Haciendo uso de la función preprogramada numpy.linalg.det.

In [ ]:
import random
import time

resultados = []

def medir_tiempo_ejecucion(func, matriz):
    inicio = time.time()
    func(matriz)
    fin = time.time()
    return fin - inicio

for i in range(2,11):
    matriz = list(np.random.rand(i,i))
    det_rec = determinante_recursivo(matriz)
    det_g = determinante_gauss(matriz)
    det_np = np.linalg.det(matriz)
    tiempo_recursivo = medir_tiempo_ejecucion(determinante_recursivo, matriz)
    tiempo_gauss = medir_tiempo_ejecucion(determinante_gauss, matriz)
    tiempo_np = medir_tiempo_ejecucion(np.linalg.det, matriz)
    resultados.append((i, tiempo_recursivo, tiempo_gauss, tiempo_np))

# Tabla
print(f"{'Tamaño':<10}{'Recursivo (s)':<20}{'Gauss (s)':<20}{'NumPy (s)':<20}")
for n, t_rec, t_gauss, t_np in resultados:
    print(f"{n:<10}{t_rec:<20.6f}{t_gauss:<20.6f}{t_np:<20.6f}")

Tamaño    Recursivo (s)       Gauss (s)           NumPy (s)           
2         0.000007            0.000024            0.000096            
3         0.000064            0.000018            0.000072            
4         0.000093            0.000026            0.000029            
5         0.000330            0.000037            0.000027            
6         0.001862            0.000057            0.000059            
7         0.009504            0.000067            0.000049            
8         0.066832            0.000115            0.000075            
9         0.618732            0.000105            0.000061            
10        5.766963            0.000124            0.000059            


In [ ]:
# Resultados

# Viendo los resultados de la tabla, podemos observar que según se aumenta el tamaño de la matriz el que menos
# tarda es numpy.linalg.det, como es lógico ya que ya está optimizada al máximo.
# También se ve de un vistazo como el método recursivo es muy impráctico, ya que es el que más tarda, porque como
# hemos dicho en el ejercicio anterior su complejidad es de O(n!), pero la de Gauss es de O(n^3) y se puede ver
# como para matrices pequeñas da un buen resultado, si es verdad que para las más grandes será mejor opción
# numpy.linalg.det, pero nuestra segunda mejor opción sería el método de Gauss.

## Ejercicio 2. Descenso del Gradiente

a) Prográmese en Python el método de descenso de gradiente para funciones de n variables. La función deberá tener como parámetros de entradas:
- El gradiente de la función que se desea minimizar ∇f (puede venir dada como otra función previamente implementada, grad f, con entrada un vector, representando el punto donde se quiere calcular el gradiente, y salida otro vector, representando el gradiente de f en dicho punto).
- Un valor inicial x0 ∈ Rn (almacenado en un vector de n componentes). El ratio de aprendizaje γ (que se asume constante para cada iteración).
- Un parámetro de tolerancia tol (con el que finalizar el proceso cuando ∥∇f(x)∥2 < tol).
- Un número máximo de iteraciones maxit (con el fin de evitar ejecuciones indefinidas en caso de divergencia o convergencia muy lenta).

La salida de la función deberá ser la aproximación del x que cumple f′(x) ≈ 0, correspondiente a la última iteración realizada en el método.

In [62]:
# Funcion principal del gradiente

import numpy as np

def descenso_gradiente(grad_f , x0 , gamma , tol , maxit):

    x = x0  # Punto inicial
    for i in range(maxit):
        gradiente = grad_f(x)  # Calcula el gradiente en el punto actual
        norma_gradiente = abs(gradiente)  # Usa la norma absoluta para una variable
        if norma_gradiente < tol:
            print(f"Convergió en {i} iteraciones. Norma del gradiente: {norma_gradiente:.6e}")
            return x
        x -= gamma * gradiente  # Actualiza el punto x
    print(f"Se alcanzó el número máximo de iteraciones ({maxit}). Norma del gradiente final: {norma_gradiente:.6e}")
    return x

b) Sea la función f: R → R dada por f(x)=3x4 +4x3 −12x2 +7.

i.  Aplica el método sobre f(x) con x0 = 3 γ = 0.001, tol=1e-12, maxit=1e5.

ii. Aplica de nuevo el método sobre f(x) con x0 = 3, γ = 0.01, tol=1e-12, maxit=1e5.

iii. Contrasta e interpreta los dos resultados obtenidos en los apartados anteriores y compáralos con los mínimos locales obtenidos analíticamente. ¿Qué influencia puede llegar a tener la elección del ratio de aprendizaje γ?

iv. Aplica nuevamente el método sobre f(x) con x0 = 3, γ = 0.1, tol=1e-12, maxit=1e5. Interpreta el resultado.

v. Finalmente, aplica el método sobre f(x) con x0 = 0, γ = 0.001, tol=1e-12,maxit=1e5. Interpreta elresultado y compáralo con el estudio analítico de f. ¿Se trata de un resultado deseable? ¿Por qué? ¿A qué se debe este fenómeno?


In [76]:
#Sea la función f: R → R dada por f(x)=3x4 +4x3 −12x2 +7
# Aplica el método sobre f(x) con x0 = 3 γ = 0.001, tol=1e-12, maxit=1e5

# Crea la funcion
def funcion(x):
    # f(x) = 3x^4 + 4x^3 - 12x^2 + 7
    return 3 * x**4 + 4 * x**3 - 12 * x**2 + 7

def derivada_funcion(x):
    # f'(x) = 12x^3 + 12x^2 - 24x
    return 12 * x**3 + 12 * x**2 - 24 * x

# Parámetros para el descenso de gradiente
punto_inicial = 3.0
tasa_aprendizaje = 0.001
tolerancia = 1e-12
iteraciones_max = int(1e5)


# Usar el método ajustado directamente con los parámetros proporcionados
resultado_polinomio = descenso_gradiente(
    grad_f=derivada_funcion,x0=punto_inicial,
    gamma=tasa_aprendizaje,tol=tolerancia,maxit=iteraciones_max)

print("Punto i. Para los parametros sobre f(x) con x0 = 3 γ = 0.001, tol=1e-12, maxit=1e5 donde f(x) = 3x^4 + 4x^3 - 12x^2 + 7" )
print(" ")
print("El resultado para el polinomio es:", resultado_polinomio)



Convergió en 831 iteraciones. Norma del gradiente: 9.912071e-13
Punto i. Para los parametros sobre f(x) con x0 = 3 γ = 0.001, tol=1e-12, maxit=1e5 donde f(x) = 3x^4 + 4x^3 - 12x^2 + 7
 
El resultado para el polinomio es: 1.0000000000000275


In [77]:
# ii)Aplica de nuevo el método sobre f(x) con x0 = 3, γ = 0.01, tol=1e-12, maxit=1e5.

punto_inicial = 3.0
tasa_aprendizaje = 0.01
tolerancia = 1e-12
iteraciones_max = int(1e5)


# Usar el método ajustado directamente con los parámetros proporcionados
resultado_polinomio = descenso_gradiente(
    grad_f=derivada_funcion,x0=punto_inicial,
    gamma=tasa_aprendizaje,tol=tolerancia,maxit=iteraciones_max)

print("Punto ii. Para los parametros sobre f(x) con x0 = 3 γ = 0.01, tol=1e-12, maxit=1e5 donde f(x) = 3x^4 + 4x^3 - 12x^2 + 7" )
print(" ")
print("El resultado para el polinomio es:", resultado_polinomio)

Convergió en 31 iteraciones. Norma del gradiente: 8.384404e-13
Punto ii. Para los parametros sobre f(x) con x0 = 3 γ = 0.01, tol=1e-12, maxit=1e5 donde f(x) = 3x^4 + 4x^3 - 12x^2 + 7
 
El resultado para el polinomio es: -1.9999999999999882


**iii) Contrasta e interpreta los dos resultados obtenidos en los apartados anteriores y compáralos con los mínimos locales obtenidos analíticamente. ¿Qué influencia puede llegar a tener la elección del ratio de aprendizaje γ?**

***Analisis***

𝛾, define el tamano de los pasos entre mas pequeno mayor precision, pero requiere de un numero mayor de iteracciones.

Como se observa en los ejercicios anteriores con un 𝛾=0.001 se utilizaron 831 iteracciones comparadas con las 31 requeridas para un 𝛾=0.01

El ratio de aprendizaje (𝛾) afecta la velocidad de convergencia y la estabilidad del método



In [78]:
# iv)Aplica nuevamente el método sobre f(x) con x0 = 3, γ = 0.1, tol=1e-12, maxit=1e5. Interpreta el resultado.

punto_inicial = 3.0
tasa_aprendizaje = 0.1 #Cuando la tasa es mas grande se vuelve divergente y genera error
tolerancia = 1e-12
iteraciones_max = int(1e5)


# Usar el método ajustado directamente con los parámetros proporcionados
resultado_polinomio = descenso_gradiente(
    grad_f=derivada_funcion,x0=punto_inicial,
    gamma=tasa_aprendizaje,tol=tolerancia,maxit=iteraciones_max)

print("Punto iv. Para los parametros sobre f(x) con x0 = 3 γ = 0.1, tol=1e-12, maxit=1e5 donde f(x) = 3x^4 + 4x^3 - 12x^2 + 7" )
print(" ")
print("El resultado para el polinomio es:", resultado_polinomio)


OverflowError: (34, 'Numerical result out of range')

**iv. Aplica nuevamente el método sobre f(x) con x0 = 3, γ = 0.1**

**Analisis**

En este caso γ es 10 veces mayor que en la interaccio de γ=0.01.

En este caso el γ es demasiado grande y al correrlo en la funcion sale error, lo que significa que el metodo empieza a oscilar alrededor del mínimo local en lugar de converger o incluso podria alejarse.




In [79]:
# v) Finalmente, aplica el método sobre f(x) con x0 = 0, γ = 0.001, tol=1e-12,maxit=1e5.

punto_inicial = 0
tasa_aprendizaje = 0.001 #Cuando la tasa es mas grande se vuelve divergente y genera error
tolerancia = 1e-12
iteraciones_max = int(1e5)

#Llamado de la funcion de descenso de gradiente
resultado_polinomio = descenso_gradiente(
    grad_f=derivada_funcion,x0=punto_inicial,
    gamma=tasa_aprendizaje,tol=tolerancia,maxit=iteraciones_max)

print("Punto v. Parametros con x0 = 0, γ = 0.001, tol=1e-12,maxit=1e5 donde f(x) = 3x^4 + 4x^3 - 12x^2 + 7" )
print(" ")
print("El resultado para el polinomio es:", resultado_polinomio)


Convergió en 0 iteraciones. Norma del gradiente: 0.000000e+00
Punto v. Parametros con x0 = 0, γ = 0.001, tol=1e-12,maxit=1e5 donde f(x) = 3x^4 + 4x^3 - 12x^2 + 7
 
El resultado para el polinomio es: 0


***Analisis para el punto V. ***

Al reemplazar x en la funcion como x vale 0, el algoritmo de descenso de gradiente detecta que se ha alcanzado un punto estacionario en la primera iteración.

La tasa de aprendizaje (γ) y la tolerancia no afectan el resultado en este caso porque el punto inicial 𝑥0=0 ya es estacionario.


***************************************************************************

c) Sea la función g : R2 → R dada por g(x,y)=x2 +y3 +3xy+1.

i. Aplíquese el método sobre g(x,y) con x0 = (−1,1), γ = 0.01, tol=1e-12, maxit=1e5.

ii. ¿Qué ocurre si ahora partimos de x0 = (0, 0)? ¿Se obtiene un resultado deseable?

iii. Realícese el estudio analítico de la función y utilícese para explicar y contrastar los resultados obtenidos en los dos apartados anteriores.

In [82]:
# i)Sea la función g : R2 → R dada por g(x,y)=x2 +y3 +3xy+1.
# Aplíquese el método sobre g(x,y) con x0 = (−1,1), γ = 0.01, tol=1e-12, maxit=1e5.

import numpy as np

def descenso_gradiente_multi(grad_g, x0, gamma, tol, maxit):

    x = np.array(x0, dtype=float)
    for i in range(maxit):
        gradiente = np.array(grad_g(x))
        norma_gradiente = np.linalg.norm(gradiente, ord=2)
        if norma_gradiente < tol:
            print(f"Convergió en {i} iteraciones. Norma del gradiente: {norma_gradiente:.6e}")
            return x
        x -= gamma * gradiente
    print(f"Se alcanzó el número máximo de iteraciones ({maxit}). Norma del gradiente final: {norma_gradiente:.6e}")
    return x


# Se crea la funcion para F(x) = x2 +y3 +3xy+1.
def funcion_g(xy):
    x, y = xy
    return x**2 + y**3 + 3*x*y + 1

# Se realiza la derivada de X y Y
def gradiente_g(xy):
    x, y = xy
    grad_x = 2 * x + 3 * y
    grad_y = 3 * y**2 + 3 * x
    return [grad_x, grad_y]

# Parámetros
punto_inicial = [-1, 1]
tasa_aprendizaje = 0.01
tolerancia = 1e-12
iteraciones_max = int(1e5)

# Aplica la funcion de desceso de gradiente
resultado = descenso_gradiente_multi(
    grad_g=gradiente_g,
    x0=punto_inicial,
    gamma=tasa_aprendizaje,
    tol=tolerancia,
    maxit=iteraciones_max
)

# Resultado final
print("Resultado para la función g(x, y):", resultado)



Convergió en 3139 iteraciones. Norma del gradiente: 9.986957e-13
Resultado para la función g(x, y): [-2.25  1.5 ]


In [83]:
# ii)ii. ¿Qué ocurre si ahora partimos de x0 = (0, 0)? ¿Se obtiene un resultado deseable?

# Parámetros
punto_inicial = [0,0]
tasa_aprendizaje = 0.01
tolerancia = 1e-12
iteraciones_max = int(1e5)

# Aplica la funcion de desceso de gradiente
resultado = descenso_gradiente_multi(
    grad_g=gradiente_g,
    x0=punto_inicial,
    gamma=tasa_aprendizaje,
    tol=tolerancia,
    maxit=iteraciones_max
)

# Resultado final
print("Resultado para la función g(x, y):", resultado)

Convergió en 0 iteraciones. Norma del gradiente: 0.000000e+00
Resultado para la función g(x, y): [0. 0.]


**Realícese el estudio analítico de la función y utilícese para explicar y contrastar los resultados obtenidos en los dos apartados anteriores.**

El gradiente en (0,0) es cero, lo que significa que (0,0) es un punto estacionario.

La funcion no proporciona un resultado deseable porque:

El método identifica(0,0) como un punto estacionario.
Este punto es un punto de silla, no un mínimo ni máximo local.